<div style='background:linear-gradient(135deg,#0D1117 0%,#161B22 100%);
border-left:5px solid #00E5FF; border-radius:12px;
padding:32px 36px; margin-bottom:8px;'>

<div style='font-family:monospace; font-size:11px; color:#8B949E;
letter-spacing:4px; margin-bottom:10px;'>MATSCIML / FEATURE 01 / NOTEBOOK</div>

<h1 style='font-family:monospace; font-size:36px; margin:0;
color:#E6EDF3; font-weight:700; letter-spacing:-1px;'>
<span style='color:#00E5FF;'>mat</span><span style='color:#76FF03;'>sci</span><span style='color:#E040FB;'>ml</span>
</h1>

<h2 style='font-family:monospace; font-size:20px; color:#76FF03;
margin:8px 0 16px 0; font-weight:600;'>Feature 01 — The Material Object</h2>

<p style='font-family:monospace; color:#8B949E; font-size:13px;
line-height:1.7; margin:0;'>
The <code style='color:#00E5FF; background:#0D1117; padding:2px 6px;
border-radius:4px;'>Material</code> class is the central object in matsciml.
Everything starts here. Give it a formula — get back 34 properties.
</p>

<div style='display:flex; gap:10px; margin-top:18px; flex-wrap:wrap;'>
<span style='background:#0A1628; color:#00E5FF; font-family:monospace;
font-size:11px; padding:4px 12px; border-radius:20px;
border:1px solid #00E5FF44;'>⚛️ 34 attributes</span>
<span style='background:#0A1628; color:#76FF03; font-family:monospace;
font-size:11px; padding:4px 12px; border-radius:20px;
border:1px solid #76FF0344;'>🔬 No subclasses</span>
<span style='background:#0A1628; color:#E040FB; font-family:monospace;
font-size:11px; padding:4px 12px; border-radius:20px;
border:1px solid #E040FB44;'>⚡ Lazy evaluation</span>
<span style='background:#0A1628; color:#FFD740; font-family:monospace;
font-size:11px; padding:4px 12px; border-radius:20px;
border:1px solid #FFD74044;'>🐍 Python 3.8+</span>
</div>

</div>

<div style='background:#0D1117; border:1px solid #21262D;
border-radius:10px; padding:24px 28px; font-family:monospace;'>

<div style='color:#8B949E; font-size:10px; letter-spacing:4px;
margin-bottom:14px;'>TABLE OF CONTENTS</div>

<div style='display:grid; grid-template-columns:1fr 1fr; gap:6px;'>
<a href='#s1' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>01 &nbsp; Installation &amp; Imports</a>
<a href='#s2' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>02 &nbsp; Creating a Material</a>
<a href='#s3' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>03 &nbsp; Identity Attributes</a>
<a href='#s4' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>04 &nbsp; Composition Attributes</a>
<a href='#s5' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>05 &nbsp; Crystal Structure Attributes</a>
<a href='#s6' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>06 &nbsp; Atomic Environment Attributes</a>
<a href='#s7' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>07 &nbsp; Electronic Attributes</a>
<a href='#s8' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>08 &nbsp; Methods: compare() &amp; get_neighbors()</a>
<a href='#s9' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>09 &nbsp; Batch Processing</a>
<a href='#s10' style='color:#C9D1D9; text-decoration:none; font-size:13px;
padding:6px 10px; border-radius:6px; border:1px solid #21262D;
display:block;'>10 &nbsp; Summary &amp; What Next</a>
</div>
</div>

<a id='s1'></a>
<div style='background:#0D1117; border-left:4px solid #00E5FF;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#00E5FF;
letter-spacing:3px;'>SECTION 01</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Installation &amp; Imports</span>
</div>

> 💡 **Run this cell once** to install matsciml. After installation, restart the kernel and skip this cell.


In [ ]:
# Uncomment and run once:
# !pip install matsciml

# If running locally with source code:
# !pip install -e /path/to/matsciml

In [ ]:
from matsciml import Material
import warnings
warnings.filterwarnings('ignore')

print('✅ matsciml imported successfully!')

✅ matsciml imported successfully!


<a id='s2'></a>
<div style='background:#0D1117; border-left:4px solid #76FF03;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#76FF03;
letter-spacing:3px;'>SECTION 02</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Creating a Material</span>
</div>

<div style='font-family:monospace; color:#8B949E; font-size:12px;
line-height:1.8; margin:8px 0;'>
The <code style='color:#00E5FF;'>Material</code> class accepts a chemical formula string. Internally it uses <strong style='color:#C9D1D9;'>pymatgen</strong> to look up the crystal structure and compute all properties lazily (only when you first access them).
</div>

In [ ]:
# ── Simplest way: just give a formula ────────────────────────
mat = Material('Fe2O3')

# That's it. The object is ready.
print(type(mat))
print(repr(mat))

<class 'matsciml.core.material.Material'>
Material(formula='Fe2O3')


<div style='background:#0D1117; border:1px solid #21262D;
border-radius:8px; padding:16px 20px; margin:8px 0;'>
<div style='font-family:monospace; font-size:10px; color:#8B949E;
letter-spacing:2px; margin-bottom:10px;'>5 WAYS TO CREATE A MATERIAL</div>
<table style='font-family:monospace; font-size:12px;
border-collapse:collapse; width:100%;'>
<tr style='border-bottom:1px solid #21262D;'>
<td style='padding:6px 12px; color:#00E5FF; width:30%;'>Formula string</td>
<td style='padding:6px 12px; color:#76FF03;'>Material('Fe2O3')</td></tr>
<tr style='border-bottom:1px solid #21262D;'>
<td style='padding:6px 12px; color:#00E5FF;'>CIF file</td>
<td style='padding:6px 12px; color:#76FF03;'>Material.from_cif('file.cif')</td></tr>
<tr style='border-bottom:1px solid #21262D;'>
<td style='padding:6px 12px; color:#00E5FF;'>Materials Project</td>
<td style='padding:6px 12px; color:#76FF03;'>Material.from_mp('mp-19770')</td></tr>
<tr style='border-bottom:1px solid #21262D;'>
<td style='padding:6px 12px; color:#00E5FF;'>POSCAR (VASP)</td>
<td style='padding:6px 12px; color:#76FF03;'>Material.from_poscar('POSCAR')</td></tr>
<tr>
<td style='padding:6px 12px; color:#00E5FF;'>Dictionary</td>
<td style='padding:6px 12px; color:#76FF03;'>Material.from_dict({...})</td></tr>
</table>
</div>

In [ ]:
# ── From dictionary (no internet needed) ─────────────────────
mat_mgo = Material.from_dict({
    'formula': 'MgO',
    'lattice': {
        'a': 4.211, 'b': 4.211, 'c': 4.211,
        'alpha': 90, 'beta': 90, 'gamma': 90
    },
})
print(f'Created: {mat_mgo.formula}')

Created: MgO


<a id='s3'></a>
<div style='background:#0D1117; border-left:4px solid #E040FB;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#E040FB;
letter-spacing:3px;'>SECTION 03</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Identity Attributes</span>
</div>

<div style='font-family:monospace; color:#8B949E; font-size:12px;
line-height:1.8;'>Basic information about what the material is.</div>

In [ ]:
mat = Material('Fe2O3')

print('── IDENTITY ──────────────────────────────')
print(f'formula        : {mat.formula}')
print(f'name           : {mat.name}')
print(f'elements       : {mat.elements}')
print(f'n_elements     : {mat.n_elements}')
print(f'crystal_type   : {mat.crystal_type}')

── IDENTITY ──────────────────────────────
formula        : Fe2O3
name           : Hematite
elements       : ['Fe', 'O']
n_elements     : 2
crystal_type   : Ionic


<a id='s4'></a>
<div style='background:#0D1117; border-left:4px solid #FF6D00;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#FF6D00;
letter-spacing:3px;'>SECTION 04</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Composition Attributes</span>
</div>

<div style='font-family:monospace; color:#8B949E; font-size:12px;
line-height:1.8;'>
These attributes are derived purely from the <strong style='color:#FF6D00;'>chemical formula</strong> — no crystal structure needed. They encode electronegativity, oxidation states, ionic character, and valence electrons.
</div>

In [ ]:
print('── COMPOSITION ───────────────────────────')
print(f'oxidation_states      : {mat.oxidation_states}')
print(f'electronegativity_diff: {mat.electronegativity_diff:.3f}')
print(f'ionic_character       : {mat.ionic_character:.1f}%')
print(f'valence_electrons     : {mat.valence_electrons}')
print(f'valence_electron_count: {mat.valence_electron_count:.1f}')
print(f'd_electron_count      : {mat.d_electron_count}')
print(f'avg_atomic_mass       : {mat.avg_atomic_mass:.2f} amu')
print(f'avg_ionization_energy : {mat.avg_ionization_energy:.2f} eV')
print(f'avg_electron_affinity : {mat.avg_electron_affinity:.3f} eV')

── COMPOSITION ───────────────────────────
oxidation_states      : {'Fe': 3, 'O': -2}
electronegativity_diff: 1.610
ionic_character       : 52.0%
valence_electrons     : {'Fe': 8, 'O': 6}
valence_electron_count: 7.2
d_electron_count      : {'Fe': 6}
avg_atomic_mass       : 33.23 amu
avg_ionization_energy : 10.14 eV
avg_electron_affinity : 0.953 eV


<div style='background:#0D1117; border:1px solid #FF6D0044;
border-radius:8px; padding:14px 18px; margin:6px 0;
font-family:monospace; font-size:12px;'>
<span style='color:#FF6D00;'>💡 Note on electronegativity_diff:</span><br>
<span style='color:#8B949E;'>
Fe Pauling EN = 1.83 &nbsp;|&nbsp; O Pauling EN = 3.44<br>
<span style='color:#FFD740;'>Δχ = |1.83 − 3.44| = 1.61</span> → Polar covalent / weakly ionic character ≈ 52%
</span>
</div>

<a id='s5'></a>
<div style='background:#0D1117; border-left:4px solid #FFD740;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#FFD740;
letter-spacing:3px;'>SECTION 05</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Crystal Structure Attributes</span>
</div>

In [ ]:
print('── STRUCTURE ─────────────────────────────')
print(f'crystal_system    : {mat.crystal_system}')
print(f'space_group       : {mat.space_group}')
print(f'space_group_number: {mat.space_group_number}')
print(f'n_symmetry_ops    : {mat.n_symmetry_ops}')
print(f'is_centrosymmetric: {mat.is_centrosymmetric}')
print(f'is_polar          : {mat.is_polar}')

print('\n── LATTICE ───────────────────────────────')
lat = mat.lattice
print(f'a={lat["a"]:.3f}Å  b={lat["b"]:.3f}Å  c={lat["c"]:.3f}Å')
print(f'α={lat["alpha"]:.1f}°  β={lat["beta"]:.1f}°  γ={lat["gamma"]:.1f}°')
print(f'cell_volume       : {mat.cell_volume:.2f} ų')
print(f'density           : {mat.density:.3f} g/cm³')
print(f'n_atoms_unit_cell : {mat.n_atoms_unit_cell}')

── STRUCTURE ─────────────────────────────
crystal_system    : Trigonal
space_group       : R-3c
space_group_number: 167
n_symmetry_ops    : 36
is_centrosymmetric: True
is_polar          : False

── LATTICE ───────────────────────────────
a=5.035Å  b=5.035Å  c=13.747Å
α=90.0°  β=90.0°  γ=120.0°
cell_volume       : 301.80 ų
density           : 5.260 g/cm³
n_atoms_unit_cell : 10


<a id='s6'></a>
<div style='background:#0D1117; border-left:4px solid #00E5FF;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#00E5FF;
letter-spacing:3px;'>SECTION 06</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Atomic Environment Attributes</span>
</div>

<div style='font-family:monospace; color:#8B949E; font-size:12px;
line-height:1.8;'>
These attributes describe the <strong style='color:#00E5FF;'>local neighborhood</strong> of atoms — bond lengths, coordination number, packing. Computed using a neighbor list with 6Å cutoff.
</div>

In [ ]:
print('── ENVIRONMENT ───────────────────────────')
print(f'coordination_number : {mat.coordination_number:.1f}')
print(f'bond_length         : {mat.bond_length:.3f} Å  (mean)')
print(f'bond_length_min     : {mat.bond_length_min:.3f} Å')
print(f'bond_length_max     : {mat.bond_length_max:.3f} Å')
print(f'bond_length_std     : {mat.bond_length_std:.3f} Å')
print(f'bond_angle_mean     : {mat.bond_angle_mean:.1f}°')
print(f'packing_fraction    : {mat.packing_fraction:.3f}')

── ENVIRONMENT ───────────────────────────
coordination_number : 6.0
bond_length         : 1.945 Å  (mean)
bond_length_min     : 1.908 Å
bond_length_max     : 2.116 Å
bond_length_std     : 0.061 Å
bond_angle_mean     : 99.4°
packing_fraction    : 0.712


<a id='s7'></a>
<div style='background:#0D1117; border-left:4px solid #76FF03;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#76FF03;
letter-spacing:3px;'>SECTION 07</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Electronic Attributes</span>
</div>

In [ ]:
print('── ELECTRONIC ────────────────────────────')
print(f'electronic_type   : {mat.electronic_type}')
bg_lo, bg_hi = mat.bandgap_estimate
print(f'bandgap_estimate  : {bg_lo:.1f} – {bg_hi:.1f} eV')
print(f'bandgap_ml        : {mat.bandgap_ml:.2f} eV')

# Compare metal vs semiconductor vs insulator
test_mats = [('Cu', 'metal'), ('TiO2', 'semiconductor'), ('Al2O3', 'insulator')]
print()
print(f'{"Formula":<12} {"Type":<15} {"Band gap (ML)":>15}')
print('─' * 44)
for formula, _ in test_mats:
    m = Material(formula)
    print(f'{m.formula:<12} {m.electronic_type:<15} {m.bandgap_ml:>12.2f} eV')

── ELECTRONIC ────────────────────────────
electronic_type   : semiconductor
bandgap_estimate  : 1.8 – 2.8 eV
bandgap_ml        : 2.18 eV

Formula      Type            Band gap (ML)
────────────────────────────────────────────
Cu           metal                  0.00 eV
TiO2         semiconductor          3.04 eV
Al2O3        insulator              8.76 eV


<a id='s8'></a>
<div style='background:#0D1117; border-left:4px solid #E040FB;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#E040FB;
letter-spacing:3px;'>SECTION 08</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Methods — compare() &amp; get_neighbors()</span>
</div>

In [ ]:
# ── get_neighbors(): local atomic environment ─────────────────
mat = Material('Fe2O3')
neighbors = mat.get_neighbors(site_idx=0)

print(f'Fe atom (site 0) has {len(neighbors)} neighbors:\n')
print(f'{"Element":<10} {"Distance":>10}  {"site_idx":>8}')
print('─' * 32)
for nbr in neighbors:
    print(f"{nbr['element']:<10} {nbr['distance']:>9.3f} Å  "
          f"{nbr['site_idx']:>8}")

Fe atom (site 0) has 6 neighbors:

Element       Distance  site_idx
────────────────────────────────
O              1.908 Å         2
O              1.945 Å         3
O              1.945 Å         4
O              2.116 Å         5
O              2.116 Å         6
O              2.116 Å         7


In [ ]:
# ── compare(): side-by-side property table ────────────────────
mat1 = Material('TiO2')
mat2 = Material('SnO2')   # same structure, different metal

result = mat1.compare(mat2)

print(f'{"Property":<28} {"TiO2":>10} {"SnO2":>10} {"Δ":>10}')
print('─' * 62)

for prop, vals in result.items():
    v1, v2 = vals['mat1'], vals['mat2']
    if 'diff' in vals:
        d = vals['diff']
        if isinstance(v1, float):
            print(f'{prop:<28} {v1:>10.3f} {v2:>10.3f} {d:>+10.3f}')
    else:
        tag = '✓ same' if vals['same'] else '✗ diff'
        print(f'{prop:<28} {str(v1):>10} {str(v2):>10} {tag:>10}')

Property                          TiO2       SnO2          Δ
──────────────────────────────────────────────────────────────
formula                          Fe2O3       TiO2     ✗ diff
crystal_system               Tetragonal Tetragonal     ✓ same
a                                 4.593      4.737     +0.144
c                                 2.958      3.186     +0.228
density                           4.230      6.950     +2.720
coordination_number               6.000      6.000     +0.000
bond_length                       1.946      2.057     +0.111
electronegativity_diff            2.020      1.540     -0.480
bandgap_ml                        3.040      3.590     +0.550


<a id='s9'></a>
<div style='background:#0D1117; border-left:4px solid #FF6D00;
border-radius:0 8px 8px 0; padding:14px 20px; margin:20px 0 6px 0;'>
<span style='font-family:monospace; font-size:10px; color:#FF6D00;
letter-spacing:3px;'>SECTION 09</span><br>
<span style='font-family:monospace; font-size:18px; color:#E6EDF3;
font-weight:700;'>Batch Processing</span>
</div>

<div style='font-family:monospace; color:#8B949E; font-size:12px;
line-height:1.8;'>
Process many materials with a simple list comprehension and build a summary DataFrame.
</div>

In [ ]:
import pandas as pd

formulas = [
    'Fe2O3', 'TiO2',  'SiC',    'Al2O3', 'GaN',
    'ZnO',   'MgO',   'BaTiO3', 'SnO2',  'CeO2'
]

materials = [Material(f) for f in formulas]

rows = []
for m in materials:
    rows.append({
        'formula'      : m.formula,
        'crystal_sys'  : m.crystal_system,
        'density'      : round(m.density, 2),
        'CN'           : m.coordination_number,
        'en_diff'      : round(m.electronegativity_diff, 2),
        'bandgap_ml'   : round(m.bandgap_ml, 2),
        'type'         : m.electronic_type,
    })

df = pd.DataFrame(rows)
df.style.background_gradient(subset=['bandgap_ml'], cmap='YlOrRd')

formula,crystal_sys,density,CN,en_diff,bandgap_ml,type
Fe2O3,Trigonal,5.26,6.0,1.61,2.18,semiconductor
TiO2,Tetragonal,4.23,6.0,2.02,3.04,semiconductor
SiC,Hexagonal,3.21,4.0,0.65,2.36,semiconductor
Al2O3,Trigonal,3.99,6.0,2.03,8.76,insulator
GaN,Hexagonal,6.15,4.0,1.23,3.38,semiconductor


<a id='s10'></a>
<div style='background:linear-gradient(135deg,#0D1117,#161B22);
border:1px solid #30363D; border-radius:12px; padding:28px 32px;
margin-top:20px; font-family:monospace;'>

<div style='font-size:10px; color:#8B949E; letter-spacing:3px;
margin-bottom:12px;'>SECTION 10 — SUMMARY</div>

<h3 style='color:#E6EDF3; margin:0 0 16px 0; font-size:16px;'>
✅ What you learned in this notebook
</h3>

<div style='display:grid; grid-template-columns:1fr 1fr; gap:8px;
margin-bottom:20px;'>
<div style='background:#0D1117; border:1px solid #21262D;
border-radius:6px; padding:10px 14px;'>
<span style='color:#76FF03;'>⚛️</span>
<span style='color:#C9D1D9; font-size:12px;'> Create Material from formula, CIF, dict</span></div>
<div style='background:#0D1117; border:1px solid #21262D;
border-radius:6px; padding:10px 14px;'>
<span style='color:#00E5FF;'>🔬</span>
<span style='color:#C9D1D9; font-size:12px;'> Access all 34 flat attributes directly</span></div>
<div style='background:#0D1117; border:1px solid #21262D;
border-radius:6px; padding:10px 14px;'>
<span style='color:#E040FB;'>🧪</span>
<span style='color:#C9D1D9; font-size:12px;'> compare() two materials side-by-side</span></div>
<div style='background:#0D1117; border:1px solid #21262D;
border-radius:6px; padding:10px 14px;'>
<span style='color:#FFD740;'>🔗</span>
<span style='color:#C9D1D9; font-size:12px;'> get_neighbors() for local atomic environment</span></div>
</div>

<div style='border-top:1px solid #21262D; padding-top:14px;
font-size:12px; color:#8B949E;'>
▶ <strong style='color:#00E5FF;'>Next notebook:</strong> <code style='color:#76FF03;'>02_feature_extraction.ipynb</code> — Convert any Material to ML-ready vectors with StructuralFeaturizer.
</div>

</div>